### ISSUE TITLE
Need SQLite versions of large queries

### ISSUE THREAD

**Issue by IanMayo** [debrief/pepys-import#756]
@arunlal-v has developed some large queries, to provide:
* search for matching datafiles
* return states, comments, contacts for set of datafiles

These have been developed/tested in PostGres.  We also need versions of them that run in SQLite.

**Comment by arunlal-v** [comment]
First draft of the procedures to extract Datafile records and subsequent Comments, Contacts, and States.

[Comments_for-v0.1.txt](https://github.com/debrief/pepys-import/files/5945228/Comments_for-v0.1.txt)
[Contacts_for-v0.1.txt](https://github.com/debrief/pepys-import/files/5945230/Contacts_for-v0.1.txt)
[States_for-v0.1.txt](https://github.com/debrief/pepys-import/files/5945233/States_for-v0.1.txt)
[Datafiles_for-v0.1.txt](https://github.com/debrief/pepys-import/files/5945235/Datafiles_for-v0.1.txt)

**Comment by arunlal-v** [comment]
Thanks @IanMayo for sharing the sample Sqlite DB file.
[pepys_training_database.db.zip](https://github.com/debrief/pepys-import/files/5958613/pepys_training_database.db.zip)

**Comment by arunlal-v** [comment]
Though the **States** table is defined with *Location* as **POINT** and *Time* as **Timestamp**, the date types of the values stored are **TEXT** and **BLOB** respectively.

```SQL
spatialite> .schema states
CREATE TABLE "States" (
        state_id CHAR(32) NOT NULL, 
        time TIMESTAMP NOT NULL, 
        sensor_id CHAR(32) NOT NULL, 
        elevation REAL, 
        heading REAL, 
        course REAL, 
        speed REAL, 
        source_id CHAR(32) NOT NULL, 
        privacy_id CHAR(32), 
        remarks TEXT, 
        created_date DATETIME, 
        "location" POINT, 
        CONSTRAINT "pk_States" PRIMARY KEY (state_id), 
        CONSTRAINT "fk_States_sensor_id_Sensors" FOREIGN KEY(sensor_id) REFERENCES "Sensors" (sensor_id) ON UPDATE cascade, 
        CONSTRAINT "fk_States_source_id_Datafiles" FOREIGN KEY(source_id) REFERENCES "Datafiles" (datafile_id) ON DELETE CASCADE ON UPDATE cascade, 
        CONSTRAINT "fk_States_privacy_id_Privacies" FOREIGN KEY(privacy_id) REFERENCES "Privacies" (privacy_id) ON UPDATE cascade
);

spatialite> select distinct typeof(time),typeof(location) from states;
text|blob
```

**Comment by arunlal-v** [comment]
@IanMayo, also kindly confirm if [CARRAY extension](https://sqlite.org/carray.html) is loaded into the Sqlite/Spatialite instance that is shipped with **Pepys/Debrief** . I'm assuming it's dependent on the host system.

**Comment by IanMayo** [comment]
When we create the SQLite table, we do request `TIMESTAMP` and `GEOMETRY`:
![image](https://user-images.githubusercontent.com/1108513/108344380-f9082c80-71d4-11eb-85fe-5d6042e5e129.png)

I guess SQLite internally chooses to use those lower level storage types.

**Comment by arunlal-v** [comment]
Yes, Ian. Sqlite (Sqlite3) supports only 5 storage classes. Extensions [would not add additional data types](https://sqlite.org/loadext.html), they only expose new functions/methods

https://stackoverflow.com/questions/8417217/sqlite-custom-datatypes

**Comment by arunlal-v** [comment]
Please find attached sqlite query for Measurements.

[sqlite-Measurements-v0.1.txt](https://github.com/debrief/pepys-import/files/6004958/sqlite-Measurements-v0.1.txt)

@IanMayo, kindly share a detailed database file if any. It'll help testing these queries. Currently, I'm inserting rows to test various scenarios.

**Comment by arunlal-v** [comment]
@IanMayo, just a design question. Should the sqlite measurement queries (States_for, Contacts_for, Comments_for) also support pagination?

**Comment by IanMayo** [comment]
Hmm.  Not to start with.  But, if performance is slow, getting them a page at a time would allow a progress monitor to increment, to show progress.

**Comment by arunlal-v** [comment]
Okay, I'll build in pagination. For now, client can set the page size as the total number of records being fetched. If required, client can reduce this on demand.

**Comment by arunlal-v** [comment]
Please find attached sqlite query for States measurements.

[sqlite-States_for-v0.2.txt](https://github.com/debrief/pepys-import/files/6012590/sqlite-States_for-v0.2.txt)

![States-for](https://user-images.githubusercontent.com/77854924/108554342-f531f280-7319-11eb-853b-e3ada303e4d1.png)

**Notes:**
#### 1) Sqlite/Spatialite doesn't support arrays or any collection data types natively. And so, the attached query had to be augmented with helper CTEs to make it consistent with PostgreSQL queries.
#### 2) As there are no arrays/collection datatypes, the Sensor_ids, Source_Ids, and Platform_Ids inputs have to be given as a single formatted string containing a list of UUIDs separated by a comma instead of array of UUIDs.

**Comment by arunlal-v** [comment]
Please find attached sqlite query for Contacts measurements.

[sqlite-Contacts_for-v0.2.txt](https://github.com/debrief/pepys-import/files/6023055/sqlite-Contacts_for-v0.2.txt)

![Contacts-for](https://user-images.githubusercontent.com/77854924/108730414-83dc8480-7551-11eb-90eb-89c3525bd663.png)

**Notes:**
#### 1) Sqlite/Spatialite doesn't support arrays or any collection data types natively. And so, the attached query had to be augmented with helper CTEs to make it consistent with PostgreSQL queries.
#### 2) As there are no arrays/collection datatypes, the Sensor_ids, Source_Ids, and Platform_Ids inputs have to be given as a single formatted string containing a list of UUIDs separated by a comma instead of array of UUIDs.
#### 3) The single parameter version ST_GeomFromText function in PostGIS accepts EWKT input 
(e.x. ```'SRID=4269;POLYGON((-125.0 65.0,-125.0 -45.0,80.0 -45.0,80.0 65.0,-125.0 65.0))'``` ), 
#### but the Spatialite version of the same accepts only WKT input (without the SRID indicator)
(e.x. ```'POLYGON((-125.0 65.0,-125.0 -45.0,80.0 -45.0,80.0 65.0,-125.0 65.0))'``` ). 
#### Hence, in the Sqlite query, the double parameter version is used which expects the WKT text input from client as the first parameter, and the second parameter (SRID 4326) is hard coded in the query itself.

**Comment by arunlal-v** [comment]
Please find attached sqlite query for Comments measurements.

[sqlite-Comments_for-v0.2.txt](https://github.com/debrief/pepys-import/files/6023325/sqlite-Comments_for-v0.2.txt)

![Comments-for](https://user-images.githubusercontent.com/77854924/108735753-ebe19980-7556-11eb-92ef-ef334627f784.png)

**Notes:**
#### 1) Sqlite/Spatialite doesn't support arrays or any collection data types natively. And so, the attached query had to be augmented with helper CTEs to make it consistent with PostgreSQL queries.
#### 2) As there are no arrays/collection datatypes, the  Source_Ids, and Platform_Ids inputs have to be given as a single formatted string containing a list of UUIDs separated by a comma instead of array of UUIDs.

**Comment by arunlal-v** [comment]
Reconcile https://github.com/debrief/pepys-import/issues/913

**Comment by IanMayo** [comment]
There is probably some outstanding work in modifying the `Pepys Import` logic in `Debrief` to use these new queries for SQLite interactions.

**Comment by IanMayo** [comment]
Defer until we understand the role of deployed analyst (w/ CF)
